domain/domain.go

```diff
@@ -433,52 +436,72 @@ func (m *MockFailure) getValue() bool {
+	mux         sync.RWMutex
+	isExpired   bool   // Whether information schema is out of date.
+	recoveredTS uint64 // It's used for recording the first txn TS of schema vaild.
+	timeInfo    struct {
+		mux            sync.RWMutex
+		lastReloadTime int64  // It's used for recording the time of last reload schema.
+		lastSuccTS     uint64 // It's used for recording the last txn TS of loading schema succeed.
+	}
+	lastSchemaVer    int64       // It's used for recording the last schema version.
 	MockReloadFailed MockFailure // It mocks reload failed.
 }
 
+func (s *schemaValidityInfo) updateSchemaVersion(version int64) {
+	atomic.StoreInt64(&s.lastSchemaVer, version)
+}
+
 func (s *schemaValidityInfo) updateTimeInfo(lastReloadTime int64, lastSuccTS uint64) {
-	s.mux.Lock()
-	defer s.mux.Unlock()
+	s.timeInfo.mux.Lock()
+	defer s.timeInfo.mux.Unlock()
 
-	s.lastReloadTime = lastReloadTime
-	s.lastSuccTS = lastSuccTS
+	s.timeInfo.lastReloadTime = lastReloadTime
+	s.timeInfo.lastSuccTS = lastSuccTS
 }
 
 func (s *schemaValidityInfo) getTimeInfo() (int64, uint64) {
-	s.mux.RLock()
-	defer s.mux.RUnlock()
+	s.timeInfo.mux.Lock()
+	defer s.timeInfo.mux.Unlock()
 
-	return s.lastReloadTime, s.lastSuccTS
+	return s.timeInfo.lastReloadTime, s.timeInfo.lastSuccTS
 }
 
-// SetValidity sets the schema validity value.
+// SetExpireInfo sets the schema validity value.
 // It's public in order to do the test.
-func (s *schemaValidityInfo) SetValidity(v bool, lastSuccTS uint64) {
+func (s *schemaValidityInfo) SetExpireInfo(expired bool, lastSuccTS uint64) {
 	s.mux.Lock()
-	if s.isValid != v {
-		log.Infof("[ddl] SetValidity, original:%v current:%v lastSuccTS:%v", s.isValid, v, lastSuccTS)
-		if !v {
-			log.Errorf("[ddl] SetValidity, schema validity is %v, lastSuccTS:%v", v, lastSuccTS)
-			s.firstValidTS = lastSuccTS
+	if s.isExpired != expired {
+		log.Infof("[ddl] SetExpireInfo, original:%v current:%v lastSuccTS:%v", s.isExpired, expired, lastSuccTS)
+		if expired {
+			log.Errorf("[ddl] SetExpireInfo, information schema is expired %v, lastSuccTS:%v", expired, lastSuccTS)
+			s.recoveredTS = lastSuccTS
 		}
-		s.isValid = v
+		s.isExpired = expired
 	}
 	s.mux.Unlock()
 }
 
-func (s *schemaValidityInfo) Check(txnTS uint64) error {
+// Check checks schema validity. It returns the current schema version and an error.
+func (s *schemaValidityInfo) Check(txnTS uint64, schemaVer int64) (int64, error) {
+	currVer := atomic.LoadInt64(&s.lastSchemaVer)
 	s.mux.RLock()
-	if s.isValid && (txnTS == 0 || txnTS > s.firstValidTS) {
+	if s.isExpired {
+		s.mux.RUnlock()
+		return currVer, ErrLoadSchemaTimeOut.Gen("InfomationSchema is out of date.")
+	}
+
+	// txnTS != 0, it means the transition isn't nil.
+	// txnTS <= s.recoveredTS, it means the transition begins before schema is recovered.
+	// schemaVer != currVer, it means the schema version is changed.
+	if txnTS != 0 && txnTS <= s.recoveredTS && schemaVer != currVer {
```

Comment:
I think `txnTS <= s.recoveredTS` condition should be removed ?

schema version change during a transaction, this transaction is invalid.